### ISSUE TITLE
Wikipedia images failing to load on Safari http://

### ISSUE THREAD

**Issue by hyanwong** [OneZoom/OZtree#454]
Works OK on chrome, but this would be good to fix. I wonder if @lentinj could possibly point me in the direction of why this is happening? Why is safari failing "access control checks" where chrome is not, and how do I dig further?

NB - this seems to work OK on safari using https.

<img width="620" alt="Screenshot 2021-12-15 at 11 31 17" src="https://user-images.githubusercontent.com/4699014/146179007-af14ff96-2dde-4c7f-8be9-f7af3a8e0dd6.png">

**Comment by lentinj** [comment]
I don't have safari handy to check, but I'm guessing what's going on is:

* You're loading an "http://onezoom.org" URL
* You fetch "https://en.wikipedia.org/api/rest_v1/page/html/Eastern%20gorilla" and plonk it into the page.
* The image URLs look like "//upload.wikimedia.org/wikipedia/commons/thumb/9/98/Mountain_gorillas_%288209001529%29.jpg", they get stuffed into the page.
* The browser tries to load the images, but as "http://", because "//" means the protocol is relative to the current page.
* Each of these images then tries to redirect to "https://"
* You go over 20 redirects because of the number of images on the page, Safari throws it's hands in the air.

I guess you could rewrite URLs to force "//" -> "https://", but I'm not sure that's a sane solution.

**Comment by lentinj** [comment]
You can see the "http://" -> "https://" image redirects in Chrome too if you poke around the network tab, but obviously isn't fazed by the situation.

**Comment by hyanwong** [comment]
Ah, that's really helpful, Thanks @lentinj ! Actually, I see I have an issue already at https://github.com/OneZoom/OZtree/issues/97 and a hack which fixes it by (as you say) forcing https (https://github.com/OneZoom/OZtree/compare/fix-wikipedia). I wonder if I should merge it?

**Comment by lentinj** [comment]
That doesn't look like it does ``src ^= "//"`` --> "https://" though, which is what you want. It's adding the origin to relative URLs.

Generally though, using jQuery to do it seems a lot more reasonable than the nightmares I had in my head. It does still feel a bit wac-a-mole, but I'm not sure you have a choice about that.

**Comment by hyanwong** [comment]
> That doesn't look like it does `src ^= "//"` --> "https://" though, which is what you want. It's adding the origin to relative URLs.

It's replacing plain http with https here:

```
https_base = base_url.replace(/^(http:\/\/|\/\/)/,"https://");
```

But you are right that it doesn't deal with "//"

> Generally though, using jQuery to do it seems a lot more reasonable than the nightmares I had in my head. It does still feel a bit wac-a-mole, but I'm not sure you have a choice about that.

Right. Sounds like this approach is worth merging though.

**Comment by hyanwong** [comment]
> > That doesn't look like it does `src ^= "//"` --> "https://" though, which is what you want. It's adding the origin to relative URLs.
> 
> It's replacing plain http with https here:
> 
> ```
> https_base = base_url.replace(/^(http:\/\/|\/\/)/,"https://");
> ```
> 
> But you are right that it doesn't deal with "//"
> 
> > Generally though, using jQuery to do it seems a lot more reasonable than the nightmares I had in my head. It does still feel a bit wac-a-mole, but I'm not sure you have a choice about that.
> 
> Right. Sounds like this approach is worth merging though.

OIC - it's not doing `src` - you're right! Sorry

**Comment by hyanwong** [comment]
Urgh - still doesn't work because wikipedia have switched to using `srcset`. This is going to be trickier than I thought. Damn.

**Comment by lentinj** [comment]
You could regex-find-and-replace within a srcset, but is keeping http://onezoom.org working worth that pain?

**Comment by hyanwong** [comment]
I think this does it:

https://github.com/OneZoom/OZtree/pull/456/commits/df5e53b095e263e7cfb5ae4b549ea3324808fc3b

**Comment by hyanwong** [comment]
Fixed & merged into 3.6
